### PR TITLE
IIBAN and custom validation

### PR DESCRIPTION
**Pull request by Elinvynia**
Addresses #1

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -22,17 +22,47 @@
...
+#[non_exhaustive]
+pub struct IBan {
+    /// The country of this IBAN.
+    pub country: Country,
+    pub(crate) raw: String,
+}
+
+impl IBan {
+    /// Returns the account number of the IBAN.
+    pub fn account_number(&self) -> u128 {
```
Comment (terahlunah) [inline src/lib.rs:49]: Why does account_number() returns an u128 and not a String ?
The BBAN are supposed to contain alphanumeric characters, not just numbers

File: src/lib.rs [hunk 2]
```diff
@@ -22,17 +22,47 @@
...
     uint::construct_uint! {
-        pub struct U256(4);
+        pub(crate) struct U256(4);
+    }
+}
+
+/// Represents an IBAN and provides helpful methods.
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct IBan {
```
Comment (terahlunah) [inline src/lib.rs]:
```suggestion
pub struct Iban{
```

File: src/lib.rs [hunk 3]
```diff
@@ -22,17 +22,47 @@
...
+    /// The country of this IBAN.
+    pub country: Country,
+    pub(crate) raw: String,
+}
+
+impl IBan {
```
Comment (terahlunah) [inline src/lib.rs]:
```suggestion
impl Iban {
```

File: src/country.rs [hunk 4]
```diff
@@ -234,6 +236,7 @@ impl Country {
             Guatemala => f!(c(4), c(20)),
             Hungary => f!(n(24)),
             Iceland => f!(n(22)),
+            Internet => f!(n(4), c(8)),
```
Comment (terahlunah) [inline src/country.rs]: The bank code of an Internet IBAN is also alpha numeric
```suggestion
            Internet => f!(c(12)),
```

File: src/country_specific.rs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+                let branch_number = &input[8..=13];
+                let mut total = 0;
+                for (ch, w) in branch_number.chars().zip([10, 5, 8, 4, 2, 1].iter()) {
+                    let ch = ch.to_digit(10).unwrap();
+                    total += ch * w;
+                }
+
+                total % 11 == 0
+            }
+            Internet => &input[0..2] == "AA",
```
Comment (terahlunah) [inline src/country_specific.rs]: This is not necessary because we know the type is Internet, so it has already matched AA

**Comment by terahlunah** [comment]
Thanks for creating the PR @Elinvynia, I made a few comments on the code

I think account_number() should return a String and not a u128 to comply with the Iban spec as BBAN can contain alphanumeric characters
That would also allow to define account_number for IIBAN
https://github.com/Elinvynia/schwifty/blob/3a300ade4c2c515b1a107d972c281b4e24df282a/src/country.rs#L290

You forgot to define bank_code() for IIBAN
https://github.com/Elinvynia/schwifty/blob/3a300ade4c2c515b1a107d972c281b4e24df282a/src/country.rs#L344

Otherwise it looks good to me

**Comment by Elinvynia** [comment]
Bank code should be defined for Internet, though I admit it gets easily lost as a lot of countries share the same position of it (but thanks to the match being exhaustive with no default, there is no risk of forgetting to add things)

**Comment by terahlunah** [comment]
Yeah, you are right, I must have missed it :)

Looks good to merge, thank you!

**Pull request merged**